### PR TITLE
[WIP][Security] Added Ldap provider

### PR DESCRIPTION
**Pull request by lyrixx**
Bug fix: no, but may be
Feature addition: yes
Backwards compatibility break: no, but may be
Symfony2 tests pass: no, not yet
Fixes the following tickets: -
Todo: Finish implementation, Add more tests. Fix \DependencyInjection\Security\Factory*`
License of the code: MIT
Documentation PR: not available yet

Hello.

First, it will be useful to have this provider into the security component. Thanks to this, the feature will be available into silex application, and also drupal 8 and ezpublish 5.

I open a PR, but my work is not finished. It could be great to have an early feedback (not about CS, just about architecture)

I read the codebase of bundles available on http://knpbundles.com/search?q=ldap but there is some issues with these bundles: 
- Dependency on `Zend\Ldap` (opensky)
- A global username / password account is not needed to check is a user is able to connect to ldap. (all bundle)
- The factory are no enough flexible.

Of course, I will rebase / squash all my commits before final merge (if happen).

I am really not sure about      f40c85b 

I tried to keep the default workflow of the user authentification
But with ldap, we can not retrieve user password. The only way to check if a user is in a ldap server, and if he is able to connect to the ldap, it is to try to connect the user against the ldap.

Then, we have an issue in `src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/UserProvider/*Factory.php`. 
For exemple in https://github.com/lyrixx/symfony/blob/1834aaa11a9342455eaf30f1c6cd7ef69f951bdb/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/FormLoginFactory.php#L65 the `$provider` (AuthentificationProvider) is hardcoded. It should not be.
So, just for my test, i changed this value (this change breaks tests).
But we have to find a way to do this cleanly. How can we do that ?
- (A) Duplicate the factory (Not ok with that)
- (B) Add a new argument createAuthProvider
- ??

With A option, there will be lot of code.
With B option, we could add BC break.

### CODE REVIEW

File: src/Symfony/Component/Security/Ldap/Ldap.php [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+                $host = 'ldaps://' . $host;
+            }
+            ldap_set_option($this->connection, LDAP_OPT_PROTOCOL_VERSION, $this->getVersion());
+            ldap_set_option($this->connection, LDAP_OPT_REFERRALS, $this->getOptReferrals());
+
+            if ($this->getUseStartTls()) {
+                ldap_start_tls($this->connection);
+            }
+
+            $this->connection = ldap_connect($host, $this->getPort());
```
Comment (ChadSikorra) [inline src/Symfony/Component/Security/Ldap/Ldap.php]: You should do the `ldap_connect()` first, then do the `ldap_set_option()` and `ldap_start_tls()`.

**Comment by jakzal** [comment]
@lyrixx if you don't have time to work on this, or there's no one to take over, I think it's better to close this PR.

I'm not fully convinced we need this in the core by the way. If the existing bundles are no good, perhaps its easier to fix them, or implement a new bundle?

**Comment by lyrixx** [comment]
@jakzal Actually, I don't need this PR anymore. So it's a bit hard for me to find time and motivation to finish it.

But I'm not sure closing this PR is our best option. I know Everybody want less open issue / PR, but This PR is almost complete. I (really) will try to finish it ! ;)

**Comment by fabpot** [comment]
... and I and it looks like a lot of people are looking for a good LDAP support in the Symfony Security Component.

**Comment by jakzal** [comment]
@lyrixx let me know if I can help with finishing it off in any way.

**Comment by csarrazi** [comment]
Two things are wrong in this PR:

* Why would we need to keep the password in the session?
* Why would we need to call the `ldap_bind()` method for each request?

We only need to

* either call `ldap_bind()` in the authenticator
* or compare the provided password with the LDAP attributes

and do this once. The only case where we would need to do this for every request is if we are using HTTP Basic as the authentication mechanism.

[Binding to an LDAP means initiating a session on the server](https://directory.apache.org/api/user-guide/2.2-binding-unbinding.html). However, there is no support in PHP for keeping alive the LDAP session (the connection is not persistent). Thus, using the `ldap_bind()` for every subsequent request also means potentially creating hundreds of LDAP sessions on the directory service. Using `ldap_bind()` is also not possible without:

* either using HTTP Basic as an authentication mechanism (which sends the password for each request)
* or keeping the user's password in the session (which would induce a security flaw)

This also means that apart from HTTP Basic, stateless authentication mechanisms won't work with the current implementation, which basically makes the user provider useless.

Until we can actually use persistent LDAP connections, or until we have a broad usage of PHP applications with WSGI-like capability, we should not call the `ldap_bind()` method in the UserProvider.

Until then, for a correct implementation, we would need:

* An LDAP user provider which simply fetches the DN from the LDAP directory, which involves a search in the directory, unless the exact mapping is already known.
* Two LDAP authentication providers, which either:
    * binds the user to the LDAP connection on first request (authentication)
    * compares the encrypted password from the LDAP information
* Loading the list of roles of the user from the LDAP attributes.

This also means that we don't need to create a new method (`loadUserByUsernameAndPassword`) in the new user provider (which currently does way more it is supposed to do).

**Comment by amenophis** [comment]
@lyrixx Should we wait for something new about this PR ?

**Comment by csarrazi** [comment]
Actually, I'm working on this PR :)

I'll give more info next week, as the code has received a major overhaul. :)

**Comment by amenophis** [comment]
Really a good news to here this !
If i can help you, don't hesitate !

**Comment by csarrazi** [comment]
Currently, I've finalized the LDAP HTTP Basic authentication provider, and finalizing the LDAP User provider (WIP).
The old

**Comment by csarrazi** [comment]
Here is the current status:

- [ ] LDAP Authentication providers
    - [x] LDAP HTTP Basic authentication provider
    - [ ] LDAP Form login authentication provider
- [x] LDAP User provider

In the updated PR, the `bind()` method is only used when we want to actually check the user's credentials. This means:

- For each request, if we are using the LDAP HTTP Basic authentication provider
- When checking the username and password, during login check, for the LDAP Form login authentication provider
- When the LDAP User provider actually needs to authenticate to the LDAP to issue queries

Regardless, as seen with @fabpot last day, I should be able to issue a WIP PR against the 2.8 branch (tests won't be working yet, though).

**Comment by amenophis** [comment]
@csarrazi Do you have news on this PR ?

**Comment by csarrazi** [comment]
Yes. The groundwork is basically done, but needs some cleanup.

Also, I'm working on an edge case, where a developer would like to use both the LDAP user provider and an LDAP authentication provider.

In the current implementation, each use a different instance of the LDAP service.
To support this case, we would need to either

* duplicate the configuration of the LDAP services (one in the authentication provider's factory, another in the user provider's factory), and have two different instances of LDAP services (and connections),
* or use the same LDAP service, but this means that the common LDAP configuration would need to be set neither in the providers node nor the firewalls section.

In short, it's basically something like:

```yml
security:
    providers:
        ldap_users:
            ldap:
                host: ldap.mydomain.tld
                uid_key: sAMAccountName
                base_dn: my_base_dn
                bind_dn: default_bind_dn
                password: P4$$w0rd
                filter: my_filter_string
    firewalls:
        demo_secured_area:
            stateless: true
            provider: ldap_users
            pattern:    ^/
            http_basic_ldap:
                host: ldap.mydomain.tld
                base: dc=mydomain,dc=local
                username_suffix: "@MYDOMAIN"
```

vs
```yml
services:
    my_ldap_service_id:
        arguments: [ ldap.mydomain.tld, my_base_dn, default_bind_dn, my_filter_string, p4$$w0rd ]

security:
    providers:
        ldap_users:
            ldap:
                id: my_ldap_service_id
                filter: my_filter_string
                uid_key: sAMAccountName
    firewalls:
        demo_secured_area:
            stateless: true
            provider: ldap_users
            pattern:    ^/
            http_basic_ldap:
                id: my_ldap_service_id
                username_suffix: "@MYDOMAIN"
```

Of course, we support both ways: if a service is provided, then we use a pre-configured service. Otherwise, we create a new service instance.

Regardless of whether we chose the a common service or two separate services, I would actually make another change, in the authentication providers: currently, all authentication providers start by fetching the user (from the user provider), before authenticating the user (call the `bind()` method). But to support this specific case, inverting the logic would prevent us to make a double `bind()` (one for the search user, another authenticating the actual user).

Does this seem okay for everyone?

**Comment by iltar** [comment]
In my opinion the configuration is supposed to be done in your user-provider and not your firewall. LDAP configuration is not related to your firewall, but just how you load the users. I'm not sure why it's even required to create "http_basic_ldap" and the form variant, what's wrong with the current implementations (or the Simple* variants for that matter), where you have to implement more than a user provider?

**Comment by csarrazi** [comment]
You don't necessarily have to use the `http_basic_ldap` or `form_login_ldap`.

These are only for authenticating against LDAP (i.e. authenticating using `ldap_bind()`, instead of comparing password hashes as, depending on your LDAP server's configuration, you may not receive the password hash in the LDAP info.

Thus, a different password validation strategy must be used with LDAP.

The LDAP user provider's goal is to actually fetch the user's information from the LDAP.

If your LDAP server returns the password hash, then you're good to go simply with the LDAP user provider (using either anonymous search, or a search user's credentials) + any "normal" firewall configuration.
However, if you don't have the password hash in the info (which is the case with most LDAP implementations), the only way to validate your credentials is by using the `ldap_bind()` PHP function.

For this, you either need:

* A new authentication provider
* implement a strategy to switch between password comparison and `ldap_bind()` in the DAO authentication provider

Also, setting the authentication provider needs to be done either by overriding `HttpBasicFactory::create()` and `FormLoginFactory:createAuthProvider()` using class inheritance, or by introducing a new option in both factories, which can be used to switch between the DAO authentication provider and LDAPBindAuthenticationProvider, and setting the default to the DAO authentication provider.

Creating new firewall factories is not necessary. But my work is based on the original PR, in which both factories were introduced (`HttpBasicLdapFactory` and `FormLoginLdapFactory`). Currently, I feel that those are not necessary, and I would actually prefer the using an option to switch between the `dao` or `ldap_bind` authentication providers.

The reason behind separating the LDAP user provider from the authentication provider is that you may use `ldap_bind()` to authenticate your users, while fetching them from another data source (entity, in_memory, etc.), or use password comparison while fetching users from an LDAP server.

**Comment by iltar** [comment]
I propose to use an implementation `SimpleAuthenticatorInterface` instead of the DAO.

**Comment by csarrazi** [comment]
Point taken. The base PR is quite old (from 2012), and the SimpleAuthenticator didn't exist at the time. But we're getting there! ;)

**Comment by csarrazi** [comment]
I've released my PR for this issue. I would be glad to receive some feedback on this.

**Comment by lyrixx** [comment]
Closing in favor of #14602

**Pull request closed without merging**